src/collector/cli.rs

```diff
@@ -18,6 +18,22 @@ pub(crate) struct CollectArgs {
     // so both the argument definition and the field population will be done manually.
     #[arg(skip)]
     pub(super) collectors: Vec<String>,
+    // Use the plural in the struct but singular for the cli parameter as we're
+    // dealing with a list here.
+    #[arg(
+        id = "probe",
+        short,
+        long,
+        help = "Add a probe on the given target. Can be used multiple times. Probes should
+follow the TYPE:TARGET pattern.
+
+Valid TYPEs:
+- kprobe: kernel probes.
+- tp: tracepoints.
```

Comment:
OK, I see your point.

>     Tool level: we do target the kernel, whatever the collector & config combinations. We might not actually install a probe in certain cases but doing so would still make sense.
>     Collect command level: here, and I'll develop later, we do target a given area. For now, networking. This includes the kernel, and potentially userspace components (which is quite important, don't get me wrong).
>     Collector level: here we do target a specific use case or data type. And this includes userspace components.

Does that mean that a collector cannot add a probe that was not added in the `collect` command? So, if you enable the `ovs` collector, or the `conntrack` collector, or the `ipsec` collector but you don't explicitly know the names of the probes and add them to the cli, they won't work?
If the answer is yes, then it's quite a burden for the user and they need to know where to look (as opposed to the collectors to encoding the knowledge of where to look for certain data).
If the answer is no, then it's not clear what collectors are affected by a global `--probe`.

Also, I'm concerned about the clarity of the UX when some probes just don't apply to some collectors. For instance, what does `packet-tracer collect --collectors=conntrack,ovs --probe kfree_skb` mean? Only the conntrack and ovs collectors are enabled and none of them collect any information from the `kfree_skb`. Will I see an empty event with no data at all on each `kfree_skb`? Will I get events from other conntrack-related or ovs-related kprobes/tracepoints?

Reply:
Could we
- Issue a warning if a probe added by the user is not used by any enabled collector
- When this happens, have a "default hook" that copies (as best as it can) the probe arguments? With the TLV event format I think it should be possible. 
?

Reply:
> 
>     * When this happens, have a "default hook" that copies (as best as it can) the probe arguments? With the TLV event format I think it should be possible.
> 
> We'll end up with addresses of struct and some values. Not sure how useful would that be, our first goal is to retrieve and make sense out of the data. Idk, I would say we don't need this, but this can still be added at some point if people ask for it.

I'm trying to make sense out of `packet-tracer collect --collectors=conntrack,ovs --probe kfree_skb` as in: "Please load the ovs and conntrack collectors because I want full info on those modules but also record kfree_skb events (just copy the args) because for some reason I find it relevant and I don't specifically need the skb module, maybe I just want to correlate them by hand or something".

Reply:
If we don't load anything and capture those events, we are left with 2 options:
- print a warning and ignore the argument
- print an error an exit
In both cases it's not a great UX

Reply:
Also, what does `packet-tracer collect --colectors=skb --probe kfree_skb --probe some_other_skb_func` mean? Skb collector will _only_ probe those targets? or it'll probe all of the targets defined by skb collector _plus_ those ones?

Reply:
> Also, what does `packet-tracer collect --colectors=skb --probe kfree_skb --probe some_other_skb_func` mean? Skb collector will _only_ probe those targets? or it'll probe all of the targets defined by skb collector _plus_ those ones?

You already answered :-) I guess those will be additional? i.e: if not already covered by the collector itself really. Correct?

Reply:
> Yep. In general I don't believe collectors will add lots of kernel probes, except for special cases like tracking. Automatically adding probes would be more the job of the profile.

That would put a burden to the user who needs to know all the names of the functions and tracepoints they want, right? I mean, not all of us are kernel developers :-)

I would expect collectors to have sane defaults, i.e: I would expect `packet-tracer --collectors=socket,skb` to generate some events on well-known socket and skb places without me having to pass a huge list of function names.


> I guess with `packet-tracer collect --collectors=conntrack,ovs --probe kfree_skb`, we are left with:
> 
>     1. Ignoring the fact that the above does not make sense.
> 
>     2. Reporting at least the parameter values. I'm not convinced this really improved the UX as the data will probably be hard to use (skb won't be tracked, other pointers won't make sense, actual values are not that common).
> 
>     3. Issue an error if no data will be reported; for this we need to validate the user defined probes with the collectors in some way.
> 
> 
> To me, option 2 would be nice to have but as a collector itself, because while it looks like something that might be useful I don't think it would improve the UX much as a fallback. It would still be confusing. This leaves us with option 3.
> 
> Having said that I also think option 1 is OKish. We can't expect magic to happen when the configuration does not. The only thing I would be worried is if this "nonsense" stays invisible. But with the common event part we would see events coming from those probes, with just the symbol name + timestamp (which isn't useful as-is, but well).
> 
> WDYT?

As a design principle I always try to create APIs that are difficult - if not impossible - to use wrongly.

Let's take 2 with the behavior implemented as a default collector. Now, let's say this collector picks up generic data (cpu_id, thread_id, com, args, timestamp, etc). This is pretty much what perf would do which is pretty cool for our baseline :-D. Besides, this probes might also be affected by filters. E.g: ` --cpu` or `--pid` filter.  In fact, let's add this collector and have it attach to anything!

So, my preference would be 2 + default collector as end goal, 1 or 3 for now.

Reply:
> Agreed, the default settings should probe a set of generic functions. In my mind this is however the job of the profiles. We can have a default one, providing what you would expect, and then targeted ones for specific use cases. If the user doesn't want that, we should also provide an option to not load a profile.

Another idea: make the probes have sane defaults so the CLI can work without profile and have an option "--no-defaults" that disables these default probes (and forces their explicit configuration). All profiles define this option. This option could even be internal-only and tied to the loading of the profile itself.

> Well, after all, we already support 2 (kind of). The common part of the event will always be reported, and it contains generic data. It's just a matter of adding a few more like cpu id, thread id, com, etc. The issue with using a dedicated collector for this is, by design, it will be possible to disable it; or we would need to have two kinds of collectors. (Supporting args retrieval might be a little too complex for this common event part thought).

Just not register it with the CLI?

Another thought: I think we already have two kinds of collectors (in mind):

- "Generic" collectors: those that collect data on any probe based on automagically detected args.
  - They accept "--probes"
  - Examples: skb, socket?, etc
- "Targeted" ones that probe a very specific set of well-known probes to extract information from small areas
  - "--probes" means nothing to these
  - Examples: : ovs, conntrack, nftables, ipsec, tcp, upd, bridge, etc.

Maybe making this difference explicit might help with our UX concerns.

> Another solution is to always have the skb module loaded, which would retrieve only basic info by default.

I don't think that's a good idea. Even if it affects a limited set of probes, the number of events could be huge on production setups and if it's not what the user is looking for, it could be a huge amount of noise.